kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/RecordConverter.java

```diff
@@ -0,0 +1,508 @@
+  }
+
+  private NameMapping createNameMapping(Table table) {
+    String nameMappingString = table.properties().get(TableProperties.DEFAULT_NAME_MAPPING);
+    return nameMappingString != null ? NameMappingParser.fromJson(nameMappingString) : null;
+  }
+
+  private Object convertValue(
+      Object value, Type type, int fieldId, SchemaUpdate.Consumer schemaUpdateConsumer) {
+    if (value == null) {
+      return null;
+    }
+    switch (type.typeId()) {
+      case STRUCT:
+        return convertStructValue(value, type.asStructType(), fieldId, schemaUpdateConsumer);
+      case LIST:
+        return convertListValue(value, type.asListType(), schemaUpdateConsumer);
+      case MAP:
+        return convertMapValue(value, type.asMapType(), schemaUpdateConsumer);
+      case INTEGER:
+        return convertInt(value);
+      case LONG:
+        return convertLong(value);
+      case FLOAT:
+        return convertFloat(value);
+      case DOUBLE:
+        return convertDouble(value);
+      case DECIMAL:
+        return convertDecimal(value, (DecimalType) type);
+      case BOOLEAN:
+        return convertBoolean(value);
+      case STRING:
+        return convertString(value);
+      case UUID:
+        return convertUUID(value);
+      case BINARY:
+      case FIXED:
+        return convertBase64Binary(value);
+      case DATE:
+        return convertDateValue(value);
+      case TIME:
+        return convertTimeValue(value);
+      case TIMESTAMP:
+        return convertTimestampValue(value, (TimestampType) type);
+    }
+    throw new UnsupportedOperationException("Unsupported type: " + type.typeId());
+  }
+
+  protected GenericRecord convertStructValue(
+      Object value,
+      StructType schema,
+      int parentFieldId,
+      SchemaUpdate.Consumer schemaUpdateConsumer) {
+    if (value instanceof Map) {
+      return convertToStruct((Map<?, ?>) value, schema, parentFieldId, schemaUpdateConsumer);
+    } else if (value instanceof Struct) {
+      return convertToStruct((Struct) value, schema, parentFieldId, schemaUpdateConsumer);
+    }
+    throw new IllegalArgumentException("Cannot convert to struct: " + value.getClass().getName());
+  }
+
+  private GenericRecord convertToStruct(
+      Map<?, ?> map,
+      StructType schema,
+      int structFieldId,
+      SchemaUpdate.Consumer schemaUpdateConsumer) {
+    GenericRecord result = GenericRecord.create(schema);
+    map.forEach(
+        (recordFieldNameObj, recordFieldValue) -> {
+          String recordFieldName = recordFieldNameObj.toString();
+          NestedField tableField = lookupStructField(recordFieldName, schema, structFieldId);
+          if (tableField == null) {
+            // add the column if schema evolution is on, otherwise skip the value,
+            // skip the add column if we can't infer the type
+            if (schemaUpdateConsumer != null) {
+              Type type = SchemaUtils.inferIcebergType(recordFieldValue, config);
```

Comment:
CMIIW but I see a little "discrepancy" here. 
When we auto create tables, [we prefer to infer the type](https://github.com/apache/iceberg/blob/f4ba90d648d1ec692134c43fc24fe389aa687d5e/kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/IcebergWriterFactory.java#L80) based on the `SinkRecord.valueSchema()`, which I think is good. 
When we evolve tables however, we infer the iceberg type based purely on the java type of the object. It would be better if we did similar logic here as the auto-create-table-code-path and gave preference to the kafka-connect-type before falling back to the java-type of the object when trying to infer the iceberg type . 

One case where this would make a difference I think is in the case of Map type objects. 
Currently, I think that if the `SinkRecord.valueSchema()` specifies a Map type for a field, we will honour that when creating a table and create the field with an iceberg MapType (relevant code [here](https://github.com/apache/iceberg/blob/19cf37edec52a39befcaa443feb5e629bd87a28e/kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SchemaUtils.java#L266)) 
However, if a new SinkRecord comes in for the same table, and this record includes a new Map field, we will evolve the table's schema to include a new StructType field instead.

Reply:
When using a schema, a struct value should never come in as a map, so this method should never get called. Also, a record will always come in as a struct, not a map, when using a schema. I'll verify that, though, and add some clarifying comments.

Reply:
I'll look into separating the converter into two, one for records with a schema and one for records without, and have a common base class. I think that might be more clear.

Reply:
I added some clarifying comments. I explored refactoring this but I felt it just added more complexity, not less.

Reply:
Ok, that was helpful 👍 

The only mildly concerning thing I see here is that in the schemaless case (i.e. `SinkRecord.valueSchema() == null`); we always convert `java.util.Map` instances into `iceStructs` (unless tables are created/evolved outside of kafka-connect). 

We've had some users of tabular/iceberg-kafka-connect raise issues about this, where they have arbitrary JSON structures coming in that are being turned into structs, resulting in schemas that are potentially thousands of fields long. 

As a result, the only reasonable solution I see for these users who wish to have `java.util.Map` interpreted as an `iceMap` (rather than `iceStruct`) automatically is for them to write an SMT that converts these `java.util.Map` instances into a `kcStruct` with a `kcSchema` which explicitly types out fields as either `kcStruct` or `kcMap` as needed. 

Curious if you'd considered this case and if you had any other thoughts on it? @bryanck 

Its unfortunate that there isn't some java runtime representation that could be used to represent the `iceStruct` case other than `java.util.Map`. Failing that, FWIW, I think defaulting to `iceStruct` when we see a `java.util.Map` in the schemaless case is the correct decision. 

--- 

kc = kafka-connect
ice = iceberg

---

<details>
  <summary>Notes for posterity</summary>

- Create table code path: 
    - If `SinkRecord.valueSchema() != null`: 
        - then the `kcSchema` returned by `SinkRecord.valueSchema()` is looked at:
            - if the `kcSchema` specifies a `kcMap`, then you get an `iceMap` in your schema
            - if the `kcSchema` specifies a `kcStruct`, then you get an `iceStruct` in your schema
    - If `SinkRecord.valueSchema() == null`: 
        - then the runtime java types are looked at for a sample record
            - if the runtime java type is a `java.util.Map` instance, then you get an `iceStruct`
            - there is no runtime java type that can be inferred as an `iceMap`

- Schema evolution code path: 
    - If `SinkRecord.getValue()` returns a `kcStruct` (i.e. in the case where `SinkRecord.valueSchema() != null`): 
        - the `kcStruct` schema is looked at:
            - if the schema specifies a `kcMap`, then you get an `iceMap`
            - if the schema specifies a `kcStruct`, then you can get an `iceStruct`
    - If `SinkRecord.getValue()` returns a `java.util.Map` instance (i.e. in the case where `SinkRecord.valueSchema() == null`): 
        - the runtime java types are looked at: 
            - if the runtime java type is a `java.util.Map` instance, then you get an `iceStruct`
            - there is no runtime java type that can be inferred as an `iceMap`
</details>

Reply:
Treating JSON objects as structs aligns with the (long standing) [PR](https://github.com/apache/kafka/pull/5001/) to add schema inference to KC. There are a couple of solutions if a user wants a map type for a field. They can use an SMT to generate a schema like you mentioned, or they can manually add the column as a map type rather than relying on auto-create and/or schema evolution to create it.

Perhaps if there is demand for it we could add an option to infer map types instead of structs for JSON objects. In general, using a schema should be strongly preferred over relying on inference, IIRC some sink connectors actually require  a schema.

Reply:
I think the main issue comes from people who have json objects with dynamic keys ("key1","key2"...) and obviously don't have a defined schema and what they really want is to store that object as a map. Having a config option like "inferJsonObjectsAsMaps" (which would infer ALL json object fields as maps) seems like a reasonable request/feature.

Reply:
Yes, it seems reasonable. The only caveat is we would need to set the value type for the map, which I think we would set to string. The converter should already support converting all types to string.

Reply:
That being said, I feel like this could be added in a follow up PR.

Reply:
> In general, using a schema should be strongly preferred over relying on inference

+10

I would have even preferred that we didn't support the schema-less approach as it's added a fair amount of complexity to the code but some users find it valuable so it seems to be worth it. 

I'll also add that I can see some performance optimizations that we can do for the schema-aware code path that are likely impossible for the schema-less path, so that may also become a reason to prefer using schema rather than relying on inference long-term. 

> Having a config option like "inferJsonObjectsAsMaps" (which would infer ALL json object fields as maps) seems like a reasonable request/feature
> The only caveat is we would need to set the value type for the map, which I think we would set to string.

If that's the caveat, I'm not so sure that this would be super valuable, and also that _seems_ like a relatively straight-forward SMT at that point? 

Also a question of if it would be worth the additional code complexity to support (users would likely want the top-level-Map returned by `SinkRecord.getValue` as an `iceStruct` but any Maps inside the top-level-Map would want to be an `iceMap` I suspect). 

Not opposed to it, just a bit iffy right now for me. 

> That being said, I feel like this could be added in a follow up PR.

For sure.